### PR TITLE
Framework: add schema for sharing/publicize

### PR DESCRIPTION
**Pull request by gwwar**
This PR adds a schema for sharing/publicize so we can avoid data shape changes as described in #3101. 

Thanks @aduth's refactors in #3417, we only need to persist connections. Other transient data like fetchingConnections will never be persisted.

## Testing Instructions
* In the console set localStorage.debug to `calypso:state`
* Make sure that your blog has some connections ( Connected to Facebook/Twitter/etc ).
* Navigate to the Editor
* Refresh the page.
* The app behaves normally and loads the persisted state/current-user from IndexedDB storage

To Validate that we can't get into an inconsistent state:
* Add a non-existent required property to `connectionsSchema`, like `foo`:
```javascript
export const connectionsSchema = {
	type: 'object',
	patternProperties: {
		'^\\d+$': {
			type: 'object',
			required: [ 'ID', 'site_ID', 'foo' ], //add a non-existent prop to test invalidations
			properties: {
				ID: { type: 'integer' },
				site_ID: { type: 'integer' },
				//...
			}
		}
	},
	additionalProperties: false
};
```
* Navigate to the Editor
* Refresh the page
* You should see a console warning that validation failed
* The Editor should load.

cc @aduth @mtias @rralian @blowery

### CODE REVIEW

File: client/state/sharing/publicize/reducer.js [hunk 1]
```diff
@@ -57,7 +65,7 @@ export function connections( state = {}, action ) {
 		case SERIALIZE:
 			return state;
 		case DESERIALIZE:
-			return state;
+			return isValidStateWithSchema( state, connectionsSchema ) ? state : {};
```
Comment (aduth) [inline client/state/sharing/publicize/reducer.js]: I'm assuming line length is what drove you to using ternary in this case and condition blocks in the others, though from a legibility perspective, I think we should be consistent (on the block?).

File: client/state/sharing/publicize/schema.js [hunk 2]
```diff
@@ -0,0 +1,55 @@
+export const fetchingConnectionsSchema = {
```
Comment (aduth) [inline client/state/sharing/publicize/schema.js]: Another consistency point, we should have a common approach to exporting schemas, since this differs from how I had named and used the exports in #3251. Differences boil down to:

```js
// #3251
export const fetchingConnections = {};
import * as schema from './schema';
isValidStateWithSchema( schema.fetchingConnections );

// #3357
export const fetchingConnectionsSchema = {};
import { fetchingConnectionsSchema } from './schema';
isValidStateWithSchema( fetchingConnectionsSchema );
```

I'd probably lean more towards your approach to enable importing specific values without concern for naming conflicts (`import as` is an alternative, albeit a bit ugly).

Another point to make is that I think we should never `export default` from schemas unless the reducer itself is not a combined reducer but rather a simple function (which is not what we're doing with [sites schema](https://github.com/Automattic/wp-calypso/blob/master/client/state/sites/schema.js)).

File: client/state/sharing/publicize/test/reducer.js [hunk 3]
```diff
@@ -48,8 +50,14 @@ describe( '#fetchingConnections()', () => {
...
 	describe( 'persistence', () => {
+		before( () => {
+			sinon.stub( console, 'warn' );
+		} );
+		after( () => {
+			console.warn.restore();
+		} );
 		it( 'should load valid persisted data', () => {
-			const persistedState = Object.freeze( {
+			const persistedState = deepFreeze( {
```
Comment (aduth) [inline client/state/sharing/publicize/test/reducer.js]: A minor point, while I think it's usually fine to sneak in small maintenance changes, it'd probably be best to do so as separate commits to keep history meaningful, and to enable a reviewer to review the commits in isolation.

File: client/state/sharing/publicize/reducer.js [hunk 4]
```diff
@@ -37,9 +43,11 @@ export function fetchingConnections( state = {}, action ) {
 		case SERIALIZE:
 			return state;
 		case DESERIALIZE:
-			return state;
+			if ( isValidStateWithSchema( state, fetchingConnectionsSchema ) ) {
```
Comment (aduth) [inline client/state/sharing/publicize/reducer.js]: I wonder how intentional we should be here about intentionally excluding items from state persistence. For example, fetching status does not seem like something that should be persisted, since it's applicable only in the context of the current browser session.

There's a potentially related discussion off of this in regards to whether fetching status makes sense to keep in the sharing tree (an otherwise "entity" tree) vs. a more disposable tree segment e.g. `ui` (related: #3337).

File: client/state/sharing/schema.js [hunk 5]
```diff
@@ -0,0 +1,19 @@
+import {
```
Comment (aduth) [inline client/state/sharing/schema.js]: Do we need this schema? When would it ever be used?

File: client/state/sharing/test/reducer.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
+/**
```
Comment (aduth) [inline client/state/sharing/test/reducer.js]: I feel this is all a bit redundant given the testing above for Publicize.

File: client/state/sharing/schema.js (hunk 5 above)
Comment (gwwar) [inline client/state/sharing/schema.js]: Use case is, the connections subtree gets thrown out because the schemas don't match. FetchingConnections has those connections marked as completed or (false). Editor then fails to load fully.

File: client/state/sharing/publicize/reducer.js (hunk 4 above)
Comment (gwwar) [inline client/state/sharing/publicize/reducer.js]: :+1: I'd be fine with not persisting this. If we choose not to persist this, I think I can ditch the top level publicize schema.

File: client/state/sharing/publicize/reducer.js (hunk 1 above)
Comment (gwwar) [inline client/state/sharing/publicize/reducer.js]: :+1: We can stick with blocks. Long ternaries tend to be more error prone.

File: client/state/sharing/publicize/schema.js (hunk 2 above)
Comment (gwwar) [inline client/state/sharing/publicize/schema.js]: I'm more inclined toward the latter too. I've opened up https://github.com/Automattic/wp-calypso/pull/3372 to fix sites naming.

File: client/state/sharing/publicize/test/reducer.js (hunk 3 above)
Comment (gwwar) [inline client/state/sharing/publicize/test/reducer.js]: :+1:

File: client/state/sharing/publicize/reducer.js (hunk 1 above)
Comment (gwwar) [inline client/state/sharing/publicize/reducer.js]: Updated in ba8b593

File: client/state/sharing/publicize/schema.js (hunk 2 above)
Comment (gwwar) [inline client/state/sharing/publicize/schema.js]: Updated in ba8b593

File: client/state/sharing/publicize/reducer.js (hunk 4 above)
Comment (gwwar) [inline client/state/sharing/publicize/reducer.js]: We no longer persist this in ba8b593

File: client/state/sharing/schema.js (hunk 5 above)
Comment (gwwar) [inline client/state/sharing/schema.js]: I backed this out in ba8b593, but we still need a way of throwing out `connectionsBySiteId` and `connectionsSchema` if one of them doesn't pass schema checks

**Comment by gwwar** [comment]
To get into an inconsistent state: 
- Start up Calypso with the testing instructions. 
- Clear browser storage. 
- Refresh the page to make sure we've persisted data. 
- Update the connectionsSchema to have another required value like `foo` which will cause data to be invalid.
- Refresh the page.
Editor fails to load

To get out of this stuck state, you'll need to revert changes to the schema, and clear browser storage.

File: client/state/sharing/schema.js (hunk 5 above)
Comment (gwwar) [inline client/state/sharing/schema.js]: I thought about this a bit, and re-added this, and made sure to wire it up this time. I'll update the PR with examples of inconsistent state.

File: client/state/sharing/test/reducer.js (hunk 6 above)
Comment (gwwar) [inline client/state/sharing/test/reducer.js]: I moved the testing for persistence one level up in 446518b

**Comment by blowery** [comment]
Perhaps `connectionsBySiteId` should be a selector instead of stored data?

**Comment by gwwar** [comment]
> Perhaps connectionsBySiteId should be a selector instead of stored data?

Interesting, I'll play around with that. I would like to avoid a larger schema file if possible.

**Comment by aduth** [comment]
>Perhaps `connectionsBySiteId` should be a selector instead of stored data? 

Ah yeah, I knew there was something about this subtree I meant to bring in line with our recommendations. We should definitely make this change.

@gwwar : Do you think it should be made separately?

**Comment by gwwar** [comment]
:bow: Thanks @aduth I'll update this PR after #3417 lands

**Comment by gwwar** [comment]
cc @aduth @rralian @blowery almost forgot about this one. It's ready for another look if you have time.

File: client/state/sharing/publicize/schema.js [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+				service: { type: 'string' },
+				label: { type: 'string' },
+				issued: { type: 'string' },
+				expires: { type: 'string' },
+				external_ID: { type: 'string' },
+				external_name: { type: [ 'string', 'null' ] },
+				external_display: { type: 'string' },
+				external_profile_picture: { type: 'string' },
+				external_profile_URL: { type: 'string' },
+				external_follower_count: { type: 'integer' },
```
Comment (aduth) [inline client/state/sharing/publicize/schema.js]: Getting a schema validation warning here. This value is `null` in my case.

image

**Comment by gwwar** [comment]
Thanks @aduth I'll take another look

**Comment by gwwar** [comment]
@aduth I've updated the schema to allow null types for all external properties.

**Comment by aduth** [comment]
Looks good :+1:

(Side-note: I'm really excited for state persistence to gain more adoption / become enabled in production :smile:  )

**Comment by gwwar** [comment]
Thanks for the review! :smile: Devs should be a bit more aware of it when #3853 enables it in stage. I'll continue to chip away at these schema PRs in my spare time.
